### DISCUSSION TITLE
Como eliminar o FTP?

### DISCUSSION THREAD

**Discussion by heliomsolivas** [frontendbr/forum#165] [Perguntas e Respostas]
Bom, o que acontece é o seguinte, no meu dia-a-dia eu recebo um projeto e tenho que voar no desenvolvimento para mostrar uma "cara" de como o projeto anda, onde eu trabalho eu controlo a versão somente no local com o git, e subo para o desenvolvimento para o cliente ver, ainda usamos o grande Filezilla e o que acontece é que na maioria das vezes alguém ou eu mesmo acaba esquecendo de subir algum arquivo. E acabo perdendo tempo caçando o arquivo ou subindo um monte de coisa novamente atoa.

Vejo pessoas falando do DeployBot e do Git mesmo, minha preferência ainda é o Git, gostaria de saber se alguém utiliza alguma outra forma de deploy, ouço muito também sobre Docker, Vagrant, etc, mas ainda não entendi muito bem na prática.

**Comment by robsonbittencourt** [comment]
@heliomsolivas existem diversas maneiras de você automatizar este processo. Por exemplo posso te mostrar o [repositório do meu blog](https://github.com/robsonbittencourt/rbittencourt.com). 

Quando faço um commit no master, com um novo post, é disparado automaticamente [um build no travis.ci](https://travis-ci.org/robsonbittencourt/rbittencourt.com). Esse build faz uma série de otimizações e após sobe os arquivos modificados para o S3 da Amazon. 

Claro tive que pesquisar para fazer isso. Mas o que posso te dizer é que você vai "perder" um tempo da primeira vez, mas após você economiza muito mais. Além de evitar esses erros que você citou.

Abaixo deixo alguns tópicos para você fazer uma pesquisa sobre esse assunto:
- Continuous Integration
- Jenkins 
- Travis CI

E qualquer coisa estamos ai para ajudar :)

**Comment by lazarofl** [comment]
Recomendo dar uma olhada no Heroku, e ate mesmo no Azure.

Pelo Heroku vc consegue subir uma aplicação gratuitamente para casos como esse que vc mencionou, tudo por linha de comando. Basicamente vc faz 'push' numa branch remota dele e a magica de deploy acontece.

No Azure vc consegue configurar pra ele escutar uma branch do github. Mas não é gratuito pelo q me lembro.

Existem N formas de fazer isso, vai depender da linguagem que vc usa. Recomendo que pesquise sobre continuous deployment platforms tb.

**Comment by rberts** [comment]
@heliomsolivas Sofro desse mesmo problema, e também estou buscando formas de fugir do ftp. Acho que a primeira coisa é buscar trabalhar com servidores que tenham acesso ssh.
Mas como a maioria do pessoal disse, isso pode variar mudo de qual linguagem/framework você trabalha.
Eu busquei algumas alternativas para o que eu costumo trabalhar no dia-a-dia.

Se você usa algum gerador de site estático, já é algo mais fácil. O próprio github já tem o gh-pages para facilitar, caso seja algum projeto privado de cliente, eu costumo usa o [gitlab](https://gitlab.com) que também tem um deploy automatizado com jekyll, e funciona um esquema de [pages](https://pages.gitlab.io) bem semelhante ao do github.

Se você trabalha com wordpress, eu comecei a usar o [wordmove](http://welaika.github.io/wordmove), que tem me ajudado bastante, ele faz um push pro server, pode ser separado em partes (all, db, theme, uploads).

E como falaram aí também tem o heroku que aceita várias linguagens, e é bem automatizado, fazendo a "mágica" com o push.

**Comment by tarciozemel** [comment]
Basicamente, isso que o @robsonbittencourt disse: Integração Contínua.

Dê uma olhada no [wercker](http://wercker.com/) -- que tem tudo a ver com o [Docker](https://www.docker.com/), ou seja, vai até melhorar seu fluxo de desenvolvimento, como um todo.

**Comment by lagden** [comment]
Eu eliminei o FTP em 2009... poderia ter eliminado antes, mas só fui conhecer colegas de auditório que manjavam de Linux pra carai nessa época!!

É sempre bom conhecer um cara bom de Linux ou infraestrutura...
Atualmente estou migrando os desenvolvimento para o [azk](http://www.azk.io/)...

O azk é brasileiro e os caras são feras!!

**Comment by theandersonn** [comment]
@heliomsolivas eu estava numa situação idêntica a sua, tentei fazer o deploy de algumas maneiras, mas não estava conseguindo colocar em pratica por não entender todos os processos. Fazia o versionamento e mandava as alterações pelo Filezilla :tired_face:

Através dessa vídeo aula e o script desenvolvido consegui configurar o servidor e fazer a festa com o Git . https://www.youtube.com/watch?v=2W6eesDo7XI

Sei que é um processo minimalista e existem outras camadas e possibilidades importantes como testes, mas para um start e entendimento do repositório e servidor, achei super válido. É a forma que estou utilizando hj.

**Comment by ghost** [comment]
@istiu Vc ainda usa o wordmove?

**Comment by joffilyfe** [comment]
Está também é uma maneira interessante de se fazer ([Artigo](http://www.akitaonrails.com/2010/02/13/deploy-com-git-push))

**Comment by raulfernando** [comment]
Eu faço desta maneira 

[Parte I](https://tableless.com.br/usando-git-para-atualizar-arquivos-no-servidor-de-hospedagem/)
[Parte II](https://tableless.com.br/usando-git-para-atualizar-arquivos-no-servidor-de-hospedagem-ii-final/)

Achei bem simples e funciona perfeito pra mim

**Comment by rberts** [comment]
@wdss93 Cara, tem um tempo que não uso. Até porque estou buscando outras opções para o wordpress (quando possível).
Porque?

**Comment by lagden** [comment]
Usem o [rsync](https://linux.die.net/man/1/rsync)

Exemplo de um script de deploy:
https://github.com/lagden/x-frame/blob/master/bin/deploy
